### ISSUE TITLE
Add support of default AWS credentials chain

### ISSUE THREAD

**Issue by chuwy** [snowplow/dataflow-runner#36]
Right now we can use:

```
"credentials": {
  "accessKeyId": "env",
  "secretAccessKey": "env"
}
```

To extract credentials from environment variables, but we also can use more default approach using `default` AWS profile etc.

**Comment by davehowell** [comment]
For use on ECS Fargate is a good reason to support this: https://andrew.hawker.io/writings/2020/02/10/snowplow-fargate-task-role/

**Comment by davehowell** [comment]
I've tried setting this up in Fargate. 
I notice that the code already supports using `iam` instead of `env` in [GetProviderCredentials()](https://github.com/snowplow/dataflow-runner/blob/master/src/utils.go#L33)

I tried using `iam` but it results in this error:

```
error: invalid memory address or nil pointer dereference
```

full error message:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x89b9d6]
goroutine 22 [running]:
github.com/aws/aws-sdk-go/aws/ec2metadata.(*EC2Metadata).GetMetadata(0x0, 0xc47452, 0x19, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/aws/ec2metadata/api.go:60 +0x146
github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds.requestCredList(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds/ec2_role_provider.go:134 +0x58
github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds.(*EC2RoleProvider).Retrieve(0xc0002e4510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds/ec2_role_provider.go:90 +0x6a
github.com/aws/aws-sdk-go/aws/credentials.(*Credentials).singleRetrieve(0xc0002e4540, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/aws/credentials/credentials.go:251 +0x29f
github.com/aws/aws-sdk-go/internal/sync/singleflight.(*Group).doCall(0xc0002e4550, 0xc0002ec7e0, 0x0, 0x0, 0xc0002d5020)
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/internal/sync/singleflight/singleflight.go:97 +0x2e
created by github.com/aws/aws-sdk-go/internal/sync/singleflight.(*Group).DoChan
	/home/travis/gopath/src/github.com/aws/aws-sdk-go/internal/sync/singleflight/singleflight.go:90 +0x2b4
```

There is some useful information in [this aws forum reply](https://forums.aws.amazon.com/thread.jspa?messageID=831798)  

> the instance metadata endpoint is not available for Fargate. Instead, you may make use of Task Metadata Endpoint...
> if you wish to make use of Task Role credentials, you may make use of the AWS_CONTAINER_CREDENTIALS_RELATIVE_URI environment variable
> the credentials URI is passed to PID 1 of the container only. So, if you are running your application as a child process, the application does not have access to the Task Role relative URI variable ...

as a workaround you can add this `export AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` in your wrapper / initd / docker-entrypoint script.

* I have done this and still seeing the error with the ec2_role_provider as above.


At this stage I have forked the repo and looking to see if I can add the default credentials support because all of the AWS documentation I have read suggest it should just work with a new session, which should default to using the credentials provider chain, which in the context of ECS/Fargate should discover the task metadata endpoint. I have a suspicion that it is the explicit use of the `ec2_role_provider` that is not working with fargate. 

If the default works then my feeling is that the `GetCredentialsProvider` function is redundant. If you want to provide environment variables to you EC2 or ECS then they take priority in the credentials provider chain, and it then discovers IAM roles, so specifying `env` or `iam` would be tautology. The other use-case, explicitly providing a hardcoded (or even templated) `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` is very bad security practice and shouldn't be provisioned for in the code anyway. 

I am happy to look at a PR for this, I've already set up the dev environment and found a few issues in the vagrant build as mentioned in https://github.com/snowplow/dataflow-runner/issues/62 which I would also include. @chuwy I would appreciate your feedback and guidance on this. For now I am making some changes and testing anyway but I'd prefer not to be running a custom fork.

**Comment by davehowell** [comment]
I've made changes and added tests that are passing, I will test the executable in my environment and then submit a PR. 

I note that two unrelated tests failed before I made any changes, and still fail. These are testing hashicorp consul related features:
1. `TestConsulLock` 
2. `TestGetLock` 

Both errors are happening in the `makeClient` function in `lock_test.go` on line 33.

```
FAIL: TestGetLock (2.02s)
    server.go:278: CONFIG JSON: {"node_name":"node-796518e7-0a25-b7b4-127b-bfb672a0efe3","node_id":"796518e7-0a25-b7b4-127b-bfb672a0efe3","performance":{"raft_multiplier":1},"bootstrap":true,"server":true,"data_dir":"/tmp/TestGetLock691941124/data","segments":null,"disable_update_check":true,"log_level":"debug","bind_addr":"127.0.0.1","addresses":{},"ports":{"dns":30427,"http":30428,"https":30429,"serf_lan":30430,"serf_wan":30431,"server":30432},"acl_enforce_version_8":false,"acl":{"tokens":{}},"connect":{"ca_config":{"cluster_id":"11111111-2222-3333-4444-555555555555"},"enabled":true}}
    io.go:404: ==> Error decoding '/tmp/TestGetLock691941124/config.json': Config has invalid keys: segments,connect,acl
    io.go:404:
    lock_test.go:33: api unavailable
```

* The testutil package mentions that it relies on using the consul cli
* The oss-playbook for the consul cli being installed is very old, version 0.8.3
* The testing depends on importing `github.com/hashicorp/consul/sdk/testutil` library
* This project was set up in the time before go modules and being able to pin dependencies. The [dark era of dependency hell](https://github.com/golang/go/wiki/PackageManagementTools)

I manually installed the latest version of consul, 1.7.2 and ran `make test` again and now all of the tests pass including these two. 
```
PASS
coverage: 72.6% of statements 
```

The fix for this will be to add a new playbook to the ansible-playbooks repo for that consul version and then update this. I had submitted a PR for that https://github.com/snowplow/ansible-playbooks/pull/135

That is not a long term fix, however. I think that this code should probably use `go mod init` and generate the dependency files.
